### DISCUSSION TITLE
feel free to edit Open3Edatapoints.py!

### DISCUSSION THREAD

**Discussion by philippoo66** [open3e/open3e#27] [General]
You are warmly welcomed to bring the project forward by adding your findings to Open3Edatapoints.py!

In order to keep everything 'nice' we considered some rules you should take into account when doing so:

- please always use English therms only
- if possible look for the terms mentioned in the English Viessmann manuals and use these. if not found, please use adequate technical terms (online sources might help)
- do not use blanks. choose writings alike "NameOfValue" or use '_' or '-' in cases like "WLAN_IP-Address"
- always start with a capital letter
- if there are unidentified parts of data in complex datapoints, capture these by RawCodec of according length and name it "Unknown"
- avoid recurrences. If the datapoint is named "FlowTemperature", its subs should be named "Actual", "Average", not "FlowTemperatureActual", "FlowTemperatureAverage"
- to commit a change please start a pull request. Automated testing will be triggered along this procedure and other developers may have a look to the changes.
![image](https://github.com/user-attachments/assets/a0fef101-0487-42af-b3a2-b3baabb4a076)
- last not least: edit only if you are reasonably sure, but also don't be hesitant too much ;-) It is always possible to change something later if found differently.

if you think, something of these 'rules' should be different, please start to discuss here below!

thank you for your great help and support!

**Comment by rechi2** [comment]
I can provide the PT2 Datapoints. But not sure how to edit them into the file. Is it possible that I send them to you?

**Reply by philippoo66** [reply]
great! sure! and then we will do it together so next time you will know how to ;-)

**Reply by philippoo66** [reply]
p.s perhaps it just requires a hook?
a.) make a fork of the open3e/develop branch in your account
b.) implement your additions/changes
c.) make a push request to merge your stuff into the project

is it this what you were not sure about?

**Comment by bomilkar** [comment]
I'm looking for a DID which provides the Current **Thermal** Output or  Current Thermal Power. Anyways the **thermal** equivalent to this:
_vcal 2488 CurrentElectricalPowerConsumptionSystem_
Can someone help?
I've been browsing through the list of DIDs but couldn't find anything.

**Reply by Wolfgang-03** [reply]
You are looking for "680_2496_CurrentThermalCapacitySystem"?

**Comment by bomilkar** [comment]
Many thanks @Wolfgang-03 !
2496 is the value I'm looking for.
However, I wonder what is behind that name **CurrentThermalCapacitySystem**
It needs some luck when picking a name. ;-)

**Comment by bomilkar** [comment]
BTW: does anybody know how "680_2496_CurrentThermalCapacitySystem" is measured/calculated?
Calculating electrical power is simple and measuring the inputs to the equation should not be difficult.
But how about thermal power?

**Comment by SanMiggel** [comment]
I would guess using
Temperature difference
Flow rate
Specific heat capacity

All of that is known


-------- Original message --------
From: Ronald Niederhagen ***@***.***>
Date: 12/3/24 10:23 (GMT+01:00)
To: open3e/open3e ***@***.***>
Cc: Subscribed ***@***.***>
Subject: Re: [open3e/open3e] feel free to edit Open3Edatapoints.py! (Discussion #27)


BTW: does anybody know how "680_2496_CurrentThermalCapacitySystem" is measured/calculated?
Calculating electrical power is simple and measuring the inputs to the equation should not be difficult.
But how about thermal power?

**Comment by bomilkar** [comment]
I agree. So how do you **measure** flow rate? The velocity profile of a given medium inside a tube depends on a number of (not so well defined) factors. Maybe an approximation is good enough?

**Reply by Philip-Wiege** [reply]
Ich bin mal so frei auf Deutsch zu antworten ;)

Ich glaube wir sind hier leider im falschen Thema gelandet. Wir können gerne ein neues Thema in den "Discussions" aufmachen. Hier ist eigentlich nur die Ankündigung von @philippoo66

Ich mach mal ein neues Thema auf.

**Comment by Dreman65** [comment]
@philippoo66  Eingangs steht '.....bring the project forward by adding your findings to Open3Edatapoints.py'. In welcher Datei hier lade ich jetzt neue Erkenntnisse zu DIDs der Vdens ab, deren DIDs z.B unterschiedliche Bytelänge zu bereits in Open3Edatapoints.py abgelegten DIDs für die Vcal haben (z.B. DID 1006 bis 1011 und 1024-1027) oder für DIDs (356, 374) eines anderen Geräts (0x688) der vdens? Bei mir am Raspi implementiere ich diese in Open3Edatapoints_680.py bzw. Open3Edatapoints_688.py, aber damit helfen ja diese findings nur mir. In der Anlage ein File mit aktuellen Beispielen.

[Open3Edatapoints_vdens_03_2025.txt](https://github.com/user-attachments/files/19556910/Open3Edatapoints_vdens_03_2025.txt)

**Reply by philippoo66** [reply]
hai Manfred @Dreman65 ,
wenn es mal wieder abweichende Datenlängen hat, haben wir ja zumin https://github.com/open3e/open3e/wiki/070-%E2%80%90-List-of-divergent-codecs als 'Zwischenlösung' eingerichtet. Da stehen schon ein paar von dir, die restlichen bitte gerne nachtragen! (is ja glaubich nur Rüberkopieren der Eintäge der txt Datei? wenn es nutzt gern mit der Zuordnung) Besten Dank!!

**Comment by bu-na** [comment]
@Dreman65 
im DID 364 der Vitodens steckt auch der Ionisationsstrom: 
        364 : O3EComplexType(6, "Flame", [O3EInt8(1, "Flame"), RawCodec(2, "Unknown"), O3EInt16(2, "IonizationCurrent", scale = 100), O3EByteVal(1,"Status")]), # Unit µA 

auch interessant:
_Open3Edatapoints_688.py
     376 : O3EComplexType(9, "MassFlowSensor", [O3EInt16(2, "CurrentValue", scale = 100), O3EInt16(2, "Min", scale=100),O3EInt16(2, "Max" ,scale=100), O3EInt16(2, "DeltaT"), O3EByteVal(1,"State")]), # Unit:  l/sec, K

**Reply by philippoo66** [reply]
@bu-na  supi, auch danke! 'hab ich rüberkopiert.

@Dreman65  auch deine txt (Inhalt) hab ich rüberkopiert, simply copy & paste.

aber nochmal: _feel free!_ to do it yourself ;-) Es ist einfach nur eine Sammlung, also keine Scheu!

**Comment by bu-na** [comment]
Noch eine Überlegung zur Textdatei:
zur Dekodierung von DID 505 und 506 ist es mit Blick auf die Visualisierung eleganter, O3ESdate(3, "Date") bzw. O3EStime(3, "Time") zu verwenden.

**Reply by philippoo66** [reply]
schön! und hast du es im Wiki ergänzt? ;-)

**Reply by Dreman65** [reply]
@bu-na  Hab ich vergessen in der Liste zu löschen :-), hab das nur bei mir so dekodiert, da dann die Weiterverarbeitung einfacher war. Steht auch in Open3Edatapoints.py nach wie vor unverändert als Datum und Zeit.
@philippoo66  Allerdings haben wir jetzt im Vitodens 300-W-Teil ein ziemliches Durcheinander. Teile dieser Liste gehören nämlich nach meinem Verständnis in die Open3Edatapoints.py. Ich werde mal versuchen, das etwas zu bereinigen, klappt aber leider erst am WE.

**Reply by Dreman65** [reply]
@philippoo66  Hi Phil, soweit alles erledigt, PullRequest für die develop Open3Edatapoints.py erstellt. Muß ich noch was tun? Wenn ja: Anleitung erforderlich :-).

**Reply by philippoo66** [reply]
einfach den Knopf drücken?! Anleitung: grüne Schaltfläche [Merge Pull Request] mit der ~rechten~ linken Maustaste klicken :-))

oder hast du keine entsprechende Berechtigung?

ps., ne, hattest du nicht, hab dir aber jetzt ne entsprechene Einladung geschickt

**Comment by Dreman65** [comment]
@philippoo66  Hi Phil, dummerweise hab ich deine Einladung angenommen und auch noch die Änderungen in Open3Edatapoints.py (development) durchgeführt. Allerdings kann ich mit der daraufhin erfolgten Fehlermeldung (Anlage) überhaupt nix anfangen. Was hab ich falsch gemacht?

[GitHub_Fehlermeldung.pdf](https://github.com/user-attachments/files/19620791/GitHub_Fehlermeldung.pdf)

**Reply by MyHomeMyData** [reply]
@Dreman65 Hier läuft ein Integrationstest auf Fehler. Die Fehlermeldung ist leider nicht sehr hilfreich.
Das Problem entsteht dadurch, dass Du zwei Unterpunkte im Datenpunkt 365 den gleichen Bezeichner gegen hast (2 x "Unkonw"). Das führt zu einem ungültigen JSON-Objekt und schließlich zu einer Fehlermeldung. Ich habe es behoben und werde den [PR](#228) gleich mergen.

Bitte grundsätzlich Änderungen nicht direkt committen, sondern **immer einen Pull Request (PR) erzeugen**. Das bietet Github beim Commit an. Dann laufen die Tests zunächst im PR und man kann Probleme beheben, bevor sie ins Repo einfließen.

**Reply by MyHomeMyData** [reply]
@philippoo66 Bitte Deinen Aufruf oben in der Diskussion um den Hinweis ergänzen, dass immer ein PR erzeugt werden soll. Am besten sollte dabei auch ein Reviewer angegeben werden, dann kann noch ein weiterer Entwickler kurz drüber schauen.

**Reply by Dreman65** [reply]
@philippoo66  Und meine Repo-Berechtigung bitte wieder löschen :-), ich mach das sicherheitshalber künftig wieder über einen Pull Request (mit Reviewer), lässt mich ruhiger schlafen :-). Danke Jürgen für den schnellen Support.

**Reply by philippoo66** [reply]
moin Manfred @Dreman65 !
sorry, ich war im Ausland und hatte mein github password nicht greifbar...
ich habe deine Berechtigung eine Stufe runtergesetzt ('triage'), damit solltest du ruhig schlafen können ;-)
Grüsse!

**Reply by philippoo66** [reply]
Moin Jürgen @MyHomeMyData !

ich dachte, das funktioniert automatisch so mit Pull Request und so?! ich meine bei mir wär das immer so gewesen, kann mich aber auch täuschen. Könntest du das bitte oben so rein schreiben wie du meinst (bevor ich in meiner Unwissenheit irgendwelchen Blödsinn schreibe...). Du hasst glaubichh die gleichen Rechte wie ich un ich kann auch _jeden_ Beitrag editieren, sollte also gehen...

**Reply by MyHomeMyData** [reply]
Moin Phil, willkommen zurück in Good Old Germany!

Habe in Deinem Eingangs-Post einen Punkt ergänzt. Wusste nicht, das ich das darf - und würde es auch nicht ohne Abstimmung tun.

Würde bei Gelegenheit gerne mal wieder telefonieren. Sag Bescheid, wenn Du wieder ganz zu Hause angekommen bist.

**Reply by philippoo66** [reply]
@MyHomeMyData so, die Seele ist hinterhergekommen. ich hatte auch schon gedacht, dich anzukingeln, bin aber wieder drüber weg gekommen...

**Comment by jsp3000k** [comment]
Hi, 
hab aktuell probleme mit meiner VX3 batterie und war auf der suche nach Daten und bin auf folgendes gestoßen.
Es sieht so aus als ob folgende datenpunkte float sind: 2990,2991,2992

ich hab mit Ai das script Open3Ecodecs.py um folgende Zeile erweitert:

```

import struct # for custom float testing
...


class O3EFloat32(udsoncan.DidCodec):
    def __init__(self, idStr: str, unit: str = ''):
        self.id = idStr
        self.unit = unit

    def decode(self, payload: bytes):
        return struct.unpack('<f', payload)[0]

    def encode(self, value):
        return struct.pack('<f', value)

    def __len__(self):
        return 4
```

und open3Edatapoints_68f.py angepasst:

```

2990 : O3EFloat32("ElectricalEnergySystemBatteryCapacityDelta"), #RawCodec(4, "ElectricalEnergySystemBatteryCapacityDelta"),
2991 : O3EFloat32("ElectricalEnergySystemBatteryCapacity", unit="ah"), #RawCodec(4, "ElectricalEnergySystemBatteryCapacity"),
2992 : O3EFloat32("ElectricalEnergySystemStateOfChargeUseable", unit="%"), #RawCodec(4, ElectricalEnergySystemStateOfChargeUseable"),
```
ElectricalEnergySystemStateOfChargeUseable scheint korrekt zu sein (aktuelle Ladung der Batterie - mindest Entladung).

Ich weiss nicht ob die Ai methode so korrekt ist (bin kein Entwickler). Könnte das sich jemand vielleicht anschauen und evtl. einbauen?

**Reply by MyHomeMyData** [reply]
Ich schau mir das an. Poste bitte zu jedem Datenpunkt einen Raw-Wert und den erwarteten gewandelten Wert zum Testen. Was hat 2990 für eine Einheit?

**Comment by MyHomeMyData** [comment]
Ich habe es in den aktuellen Feature Branch `feat_device_switch` integriert. Testen mit
```
pip install --upgrade git+https://github.com/open3e/open3e.git@feat_device_switch
```
Die Anzahl der Nachkommastellen kann in der Datenpunktdefinition durch `decimals=x` festgelegt werden. Default sind 2 Nachkommastellen.

**Reply by philippoo66** [reply]
oh, Viessmann hat IEEE entdeckt?

**Comment by Dreman65** [comment]
@philippoo66 @MyHomeMyData Hallo Phil, hallo Jürgen, ich hab mich mal wider besseres Wissens und trotz schlechter Erfahrungen dazu durchgerungen einen Pullrequest für die Open3Edatapoints.py in Github zu starten. Hab also wie oben beschrieben einem neuen Branch erzeugt, mit dem develop Branch synchronisiert und die entsprechenden Änderungen in Open3Edatapoints.py, Open3EdatapointsVariants.py und Open4Eenums.py durchgeführt und alle Änderungen übernommen. Dabei wurde der erste Durchlauf Open3Edatapoints.py mit einer Fehlermeldung abgeschlossen: trotz aller Sorgfalt war in einer Definition ein " zuviel :-( . Nach Behebung des Fehlers erfolgte der 2. Durchlauf fehlerfrei. Ich erhalte jetzt vor Starten des Pullrequests folgenden Status in Github:

<img width="997" height="823" alt="Pullrequest1" src="https://github.com/user-attachments/assets/4fcc94b7-ae31-409b-a12b-c3500f1b63be" />
<img width="991" height="475" alt="Pullrequest2" src="https://github.com/user-attachments/assets/a8e85cb8-569c-4c5f-a58f-3bc96478fec0" />

Dabei verwirrt mich, dass das Ergebis des ersten **fehlerhaften** Durchlaufs (16 hrs ago) auch mit angezeigt wird und ich keine Möglichkeit finde, dies zu löschen. Oder ist das garnicht notwendig und ich kann den PullRequest starten? Oder brauche ich einen von euch beiden als 'Tatortreiniger' :-)? Danke im Voraus für die Hilfe.

**Reply by MyHomeMyData** [reply]
@Dreman65 Hallo Manfred, Du hast 2 Commits erzeugt und die werden Dir angezeigt. So funktioniert Github und da sollte man ohne wichtigen Grund nicht dran rum schrauben. Der erste Commit hat einen Fehler geworfen, den Du mit dem nächsten Commit behoben hast. Genau das ist die übliche Arbeitsweise. So bleibt auch nachvollziehbar, was zur Fehlerbehebung geändert wurde.

Wenn Du den PR für open3e erzeugt, wähle aber bitte den Branch `develop` als Ziel aus. Im Screenshot hast Du `master` ausgewählt.

**Reply by MyHomeMyData** [reply]
@Dreman65 Noch ein Hinweis: Wenn man in open3e Änderungen an den Datenpunktfiles (.py) committet, wird automatisch eine Github-Action ausgeführt, die die korrespondierenden .json- und .md-Files neu erzeugt und committet. Man muss also nach Abschluss der Action nochmal ein `git pull` machen.
